### PR TITLE
utils/kdbg: Use procfs to support cpuload, irqinfo, ps, and stackmonitor

### PR DESCRIPTION
**Pull request by jeongarmy**
User can NOT call functions of kernel and use resources in kernel side.
So kdbg uses procfs to get system information such as memory, task/pthread, interrupts.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [aad7e5db65c541db40ed1daca10ed637d6832329] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [aad7e5db65c541db40ed1daca10ed637d6832329] - Code Rule Check (C++) OK.

File: apps/system/utils/Kconfig [hunk 1]
```diff
@@ -53,7 +53,7 @@ endif #NET_CMDS
 config ENABLE_CPULOAD_CMD
 	bool "cpuload"
 	default y
-	depends on SCHED_CPULOAD
+	depends on SCHED_CPULOAD && FS_PROCFS && !_FS_PROCFS_EXCLUDE_CPULOAD
```
Comment (sunghan-chang) [inline apps/system/utils/Kconfig]: 1. Please remove `_` at start of  `_FS_PROCFS_EXCLUDE_PROCESS`.
2. Is there a way to enable `FS_PROCFS_EXCLUDE_PROCESS` without enabling `FS_PROCFS`?
   If there is not, let's remove `FS_PROCFS` from dependancy. It's a redundant.

File: apps/system/utils/kdbg_utils.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_utils.c]: year

File: apps/system/utils/kdbg_utils.h [hunk 3]
```diff
@@ -38,4 +39,31 @@ struct stkmon_save_s {
 };
 #endif
 
+#if defined(CONFIG_ENABLE_IRQINFO_CMD) || defined(CONFIG_ENABLE_IRQINFO_CMD) || defined(CONFIG_ENABLE_PS_CMD) || defined(CONFIG_ENABLE_STACKMONITOR_CMD)
+
+#define PROC_MOUNTPOINT "/proc"
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_utils.h]: Let's tidy up this.
1. If you want to fix a name of proc mount point, let's move this at some public header.
2. If you want to give an option for it, let's make a new config and use it at all of this kind definition.

File: apps/system/utils/kdbg_utils.h [hunk 4]
```diff
@@ -38,4 +39,31 @@ struct stkmon_save_s {
...
+
+enum proc_stat_entry_e {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_utils.h]: what's the meaning of `proc stat entry`?

File: apps/system/utils/kdbg_utils.h [hunk 5]
```diff
@@ -38,4 +39,31 @@ struct stkmon_save_s {
...
+	PROC_STAT_PEAKHEAP,
+#if CONFIG_TASK_NAME_SIZE > 0
+	PROC_STAT_NAME,
+#endif
+	PROC_STAT_MAX,
+};
+
+typedef char *stat_entry;
+typedef int (*procentry_handler_t)(struct dirent *, void *);
+typedef void (*kdbg_parse)(char *);
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_utils.h]: Please fill argument names. That's our coding rules.

File: os/fs/procfs/fs_procfsproc.c [hunk 6]
```diff
@@ -103,8 +104,12 @@
  * accessed via the procfs file system.
  */
 
+/* '/proc/[pid]/stat' and '/proc/[pid]/status' entries are almost similar, they provide information about task/pthread.
+ * The 'status' entry is easier for humans to parse.
+ */
 enum proc_node_e {
 	PROC_LEVEL0 = 0,			/* The top-level directory */
+	PROC_STAT,					/* Task/thread status vales */
```
Comment (sunghan-chang) [inline os/fs/procfs/fs_procfsproc.c]: 1. Let's align comments.
2. What is a WORD difference between `STAT` and `STATUS`?
3. What is a difference between `status values` and `status`?

File: apps/system/utils/kdbg_ps.c [hunk 7]
```diff
@@ -77,26 +84,71 @@ static const char *kdbg_ttypenames[4] = {
...
+		stat_info[i] = strtok(NULL, " ");
+	}
+
+	flags = atoi(stat_info[PROC_STAT_FLAG]);
+	if (flags < 0) {
+		return;
+	}
+
+	state = atoi(stat_info[PROC_STAT_STATE]);
+	if (state < 0) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_ps.c]: When `stat_info[XXX]` is null, atoi returns 0. I think you should add equal.

File: apps/system/utils/kdbg_ps.c [hunk 8]
```diff
@@ -77,26 +84,71 @@ static const char *kdbg_ttypenames[4] = {
 	"--?--  "
 };
 
-static void kdbg_pseach(FAR struct tcb_s *tcb, FAR void *arg)
+static void kdbg_parse_ps(char *buf)
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_ps.c]: Is main purpose of this function `parse`? I think it's a printing.

**Comment by sunghan-chang** [comment]
@jeongarmy Please check build failure. You missed necessary headers.

**Comment by sunghan-chang** [comment]
@jeongarmy Let's rewrite a commit title because I can't know what you did through this commit.
i.e.
```
apps/utils: use procfs to support cpuload, irqinfo, ps, stackmonitor
```

File: os/fs/procfs/fs_procfsproc.c (hunk 6 above)
Comment (jeongarmy) [reply os/fs/procfs/fs_procfsproc.c]: We support 2 kinds of entry to provide information of task/pthread.
The difference is a form of printed string only.
The 'status' entry is easier for human to parse.

**Comment by seinfra** [comment]
Target : [dc136a82f45c3f5da67e3ece508238129231a044] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0f04ee7032c90252503ac663ad2c83739f7537df] -

**Comment by seinfra** [comment]
Target : [d7c312e58a17021ca30e5f697abeac0ad9ca0860] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8e4dd7d1456d83f822294388e67084ea9bf6f948] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8e4dd7d1456d83f822294388e67084ea9bf6f948] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1c1354fc8a540645f3a5ce6524c8518b6aa42369] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1c1354fc8a540645f3a5ce6524c8518b6aa42369] - Code Rule Check (C++) OK.

File: apps/system/utils/kdbg_utils.h [hunk 9]
```diff
@@ -38,4 +39,32 @@ struct stkmon_save_s {
...
+	PROC_STAT_PEAKHEAP,
+#if CONFIG_TASK_NAME_SIZE > 0
+	PROC_STAT_NAME,
+#endif
+	PROC_STAT_MAX,
+};
+
+typedef char *stat_data;
+typedef int (*procentry_handler_t)(FAR struct dirent *entryp, FAR void *arg);
+typedef void (*kdbg_print)(char *buf);
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_utils.h]: Let's add `_t` postfix.

File: os/include/tinyara/fs/fs.h [hunk 10]
```diff
@@ -86,6 +86,11 @@
 #define CONFIG_MOUNT_POINT "/mnt/"
 #endif
 
+#ifdef CONFIG_FS_PROCFS
```
Comment (sunghan-chang) [inline os/include/tinyara/fs/fs.h]: Please split this change as a different commit.

File: apps/system/utils/kdbg_stackmonitor.c [hunk 11]
```diff
@@ -134,50 +139,84 @@ static void stkmon_inactive_check(void)
...
+
+	return OK;
+}
+
 static void *stackmonitor_daemon(void *arg)
 {
+	int ret;
+
+	/* Mount Procfs to use */
+	ret = mount(NULL, PROCFS_MOUNT_POINT, PROCFS_FSTYPE, 0, NULL);
```
Comment (jeongchanKim) [inline apps/system/utils/kdbg_stackmonitor.c]: I think you'd better move mount to kdbg_stackmonitor().
This is a daemon which runs periodically, but there is not needed to mount for every time.

File: os/fs/fs_initialize.c [hunk 12]
```diff
@@ -91,7 +91,7 @@ void fs_auto_mount(void)
 		const char *fs_mountpoint;
 	} fs_automount[] = {
 #ifdef CONFIG_FS_AUTOMOUNT_PROCFS
-		{"procfs", "/proc"},
+		{PROCFS_FSTYPE, PROCFS_MOUNT_POINT},
```
Comment (sunghan-chang) [inline os/fs/fs_initialize.c]: header?

File: apps/system/utils/Kconfig [hunk 13]
```diff
@@ -53,7 +53,7 @@ endif #NET_CMDS
 config ENABLE_CPULOAD_CMD
 	bool "cpuload"
 	default y
-	depends on SCHED_CPULOAD
+	depends on SCHED_CPULOAD && FS_PROCFS && !FS_PROCFS_EXCLUDE_CPULOAD
```
Comment (sunghan-chang) [inline apps/system/utils/Kconfig]: Give me an answer for previous question.

File: apps/system/utils/kdbg_cpuload.c [hunk 14]
```diff
@@ -19,17 +19,23 @@
...
 #include <tinyara/config.h>
+#include <sched.h>
 #include <stdio.h>
 #include <string.h>
-#include <sched.h>
+#include <stdlib.h>
 #include <stdint.h>
 #include <stdbool.h>
+#include <dirent.h>
 #ifdef CONFIG_ENABLE_CPULOAD_MONITOR
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_cpuload.c]: Is it possible to compile this file without enabling this config?

File: apps/system/utils/kdbg_irqinfo.c [hunk 15]
```diff
@@ -15,19 +15,42 @@
...
 int kdbg_irqinfo(int argc, char **args)
 {
-#ifdef CONFIG_DEBUG_IRQ_INFO
-	irq_info();
+	int ret;
+	char *filepath;
+	char buf[IRQ_BUFLEN];
+
+	/* Mount Procfs to use */
+	ret = mount(NULL, PROCFS_MOUNT_POINT, PROCFS_FSTYPE, 0, NULL);
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_irqinfo.c]: Why do you add this even you modified fs_initialize.c file for fs_auto_mount()?

File: apps/system/utils/kdbg_ps.c [hunk 16]
```diff
@@ -77,26 +87,83 @@ static const char *kdbg_ttypenames[4] = {
...
+
+#if (CONFIG_TASK_NAME_SIZE > 0)
+	printf(" | %s\n", stat_info[PROC_STAT_NAME]);
+#else
 	printf("\n");
+#endif
+
+}
+
+static int ps_read_pid(FAR struct dirent *entryp, FAR void *arg)
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_ps.c]: read_pid? read_proc?

File: build/configs/qemu/smartfs/defconfig [hunk 17]
```diff
@@ -546,8 +546,21 @@ CONFIG_SMARTFS_ALIGNED_ACCESS=y
 # CONFIG_SMARTFS_DYNAMIC_HEADER is not set
 # CONFIG_SMARTFS_JOURNALING is not set
 # CONFIG_SMARTFS_SECTOR_RECOVERY is not set
-# CONFIG_FS_PROCFS is not set
+CONFIG_FS_PROCFS=y
```
Comment (sunghan-chang) [inline build/configs/qemu/smartfs/defconfig]: Let's use an another commit to push defconfig changes

File: os/fs/procfs/fs_procfsproc.c (hunk 6 above)
Comment (sunghan-chang) [reply os/fs/procfs/fs_procfsproc.c]: I don't understand. Please let me clarify this difference.

File: os/fs/fs_initialize.c (hunk 12 above)
Comment (jeongarmy) [reply os/fs/fs_initialize.c]: Done

File: os/include/tinyara/fs/fs.h (hunk 10 above)
Comment (jeongarmy) [reply os/include/tinyara/fs/fs.h]: Done.

File: apps/system/utils/kdbg_cpuload.c (hunk 14 above)
Comment (jeongarmy) [reply apps/system/utils/kdbg_cpuload.c]: yes. I checked

File: apps/system/utils/kdbg_irqinfo.c (hunk 15 above)
Comment (jeongarmy) [reply apps/system/utils/kdbg_irqinfo.c]: I added for making all use same path when automount is enabled

**Comment by seinfra** [comment]
Target : [26bd0a5bdf64b4c29592dda83908ba7b010db271] - Code Rule Check OK.

File: apps/system/utils/kdbg_stackmonitor.c (hunk 11 above)
Comment (jeongarmy) [reply apps/system/utils/kdbg_stackmonitor.c]: I think it is right because mount is out of while loop.

File: os/fs/procfs/fs_procfsproc.c (hunk 6 above)
Comment (jeongarmy) [reply os/fs/procfs/fs_procfsproc.c]: The 'stat' provides system information values only.
On the other hand, 'status' shows these values with prefix about info in details.

**Comment by seinfra** [comment]
Target : [26bd0a5bdf64b4c29592dda83908ba7b010db271] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e3fb5c04f912ce3f084447327027d4872919d633] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e3fb5c04f912ce3f084447327027d4872919d633] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e9683980cb18d24ead91b3d7d5cfc0e68ee22061] -

**Comment by seinfra** [comment]
Target : [e9683980cb18d24ead91b3d7d5cfc0e68ee22061] -

**Comment by seinfra** [comment]
Target : [b6c2fbed983508d470c6ae9e68ab15728c780c40] -

**Comment by seinfra** [comment]
Target : [272800be20783fa9c9f377b74b6cc483b83f32f6] -

**Comment by seinfra** [comment]
Target : [1dd27d18dc544e3f9e6e0e24b21e7e57fae23796] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1dd27d18dc544e3f9e6e0e24b21e7e57fae23796] -

**Comment by seinfra** [comment]
Target : [6065288bc944f52ebf0b82e717544d87e17bcb3f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6065288bc944f52ebf0b82e717544d87e17bcb3f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8b1f7dc003c09af8bd7dc71e140d53ae822b7e01] - Code Rule Check OK.
